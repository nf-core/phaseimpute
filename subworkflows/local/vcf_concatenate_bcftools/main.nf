include { BCFTOOLS_CONCAT } from '../../../modules/nf-core/bcftools/concat'

workflow VCF_CONCATENATE_BCFTOOLS {

    take:
    ch_vcf_tbi // channel: [ [id, panel, chr, tools], vcf, tbi ]

    main:

    ch_versions = Channel.empty()

    // Keep only id from meta
    ch_vcf_tbi_grouped = ch_vcf_tbi
        .map{ metaIPTC, vcf, tbi -> [metaIPTC.subMap("id", "tools", "panel") + ["chr": "all"], vcf, tbi] }
        .groupTuple( by:0 )

    // Ligate and concatenate chunks
    BCFTOOLS_CONCAT(ch_vcf_tbi_grouped)
    ch_versions = ch_versions.mix(BCFTOOLS_CONCAT.out.versions.first())

    // Join VCFs and TBIs
    ch_vcf_tbi_join = BCFTOOLS_CONCAT.out.vcf
        .join(BCFTOOLS_CONCAT.out.tbi)

    emit:
    vcf_tbi      = ch_vcf_tbi_join // channel: [ [id], vcf, tbi ]
    versions     = ch_versions     // channel: [ versions.yml ]
}
