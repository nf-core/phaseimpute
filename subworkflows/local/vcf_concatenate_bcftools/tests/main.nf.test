nextflow_workflow {

    name "Test Subworkflow VCF_CONCATENATE_BCFTOOLS"
    script "../main.nf"

    config "./nextflow.config"

    workflow "VCF_CONCATENATE_BCFTOOLS"

    tag "subworkflows"
    tag "subworkflows_local"
    tag "subworkflows/vcf_concatenate_bcftools"
    tag "vcf_concatenate_bcftools"

    tag "bcftools"
    tag "bcftools/concatenate"


    test("Concatenate vcf") {
        when {
            params {
                max_cpus   = 2
                max_memory = '2.GB'
            }
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [id: "1000GP", chr: "chr22"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr22/1000GP.chr22.s.norel.vcf.gz.csi", checkIfExist:true),
                    ],
                    [
                        [id: "1000GP", chr: "chr21"],
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz", checkIfExist:true),
                        file(params.pipelines_testdata_base_path + "hum_data/panel/chr21/1000GP.chr21.s.norel.vcf.gz.csi", checkIfExist:true),
                    ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.vcf_tbi.collect{
                    path(it[1]).vcf.summary
                    }).match("Concat content")
                }
            )
        }
    }
}
