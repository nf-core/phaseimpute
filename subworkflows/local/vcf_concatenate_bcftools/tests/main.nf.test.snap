{
    "Concat content": {
        "content": [
            [
                "VcfFile [chromosomes=[chr21, chr22], sampleCount=3196, variantCount=1739, phased=true, phasedAutodetect=true]"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T11:56:39.376973769"
    },
    "Concatenate vcf": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "all"
                        },
                        "1000GP_all.vcf.gz:md5,9836dd74da2ddd7956adf3e4f83f80e1",
                        "1000GP_all.vcf.gz.tbi:md5,e7180bb953d2bd657c420a5f76a7164d"
                    ]
                ],
                "1": [
                    "versions.yml:md5,0370775d4b565f1cf95dcbea437fa36a"
                ],
                "vcf_tbi": [
                    [
                        {
                            "id": "1000GP",
                            "chr": "all"
                        },
                        "1000GP_all.vcf.gz:md5,9836dd74da2ddd7956adf3e4f83f80e1",
                        "1000GP_all.vcf.gz.tbi:md5,e7180bb953d2bd657c420a5f76a7164d"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0370775d4b565f1cf95dcbea437fa36a"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-07-18T11:56:36.85523481"
    }
}